File: slurmemu/src/copper.rs
//
//	Emulate Copper
//

enum CopperState {
	Idle,
	Begin,
	Execute,
	WaitV,
	WaitH,
	WriteReg,
}

pub struct Copper {
	enable : bool,
	copper_list: [u16; 512],
	copper_pc: u16,
	copper_state: CopperState,
	r : u8,
	g : u8,
	b : u8,
	reg_addr : u16,
	hv_wait : u16,
	y_flip: u16,
	y_flip_en: bool,
	x_pan: u16,
	global_alpha: u8,
	alpha_override : bool
}

impl Copper {

	pub fn new() -> Self {
		Copper { 
			enable : false,
			copper_list: [0 as u16; 512],
			copper_pc: 0 as u16,
			copper_state: CopperState::Idle,
			r 		: 255,
			g 		: 0,
			b 		: 0,
			reg_addr 	: 0,
			hv_wait 	: 0,
			y_flip 		: 0,
			y_flip_en 	: false,
			x_pan 		: 0,
			global_alpha 	: 15,
			alpha_override  : true
		}
	}

	pub fn get_alpha(&mut self) -> (bool, u8)
	{
		(self.alpha_override, self.global_alpha)
	}

	pub fn write_mem(& mut self, addr: u16, data: u16)
	{
		self.copper_list[addr as usize] = data;
	}

	pub fn begin(& mut self)
	{
		// Wait state
		self.copper_state = CopperState::Execute;
	}

	pub fn port_op(& mut self, port : u16, val: u16, write : bool) -> u16
	{
		if write {
			match port & 0xf {
				0 => /* copper control */ { self.enable = val & 1 == 1; },
				1 => /* copper Y flip */ {
					self.y_flip = val & 0x1ff;
					self.y_flip_en = (val & 0x8000) == 0x8000;
				},
				2 => /* background color */ {
					self.r = ((val & 0xf00) >> 4) as u8;
					self.g = (val & 0x0f0) as u8;
					self.b = ((val & 0x00f) << 4) as u8;
				},
				3 => /* x pan */ {
					self.x_pan = val;
				},
				4 => /* alpha */ {
					self.alpha_override = (val & 0x8000) == 0x8000;
					self.global_alpha   = (val & 0xf) as u8;
				},
				_ => {},
			}
		}

		0
	}

	pub fn execute(& mut self, x : u16, y : u16)
	{
		let mut inc_pc : bool = true;

		let ins = self.copper_list[(self.copper_pc & 0x1ff) as usize];

		match (ins & 0xf000) >> 12 { 
			0x1 => /* jump */
			{
				inc_pc = false;
				self.copper_pc = ins & 0x1ff;
			},
			0x2 => /* V wait */
			{
				self.hv_wait = ins & 0x3ff;
				self.copper_state = CopperState::WaitV;
			},
			0x3 => /* H wait */
			{
				self.hv_wait = ins & 0x3ff;
				self.copper_state = CopperState::WaitH; 
			},
			0x4 => /* skip V */
			{
				if y >= ins & 0x3ff {
					self.copper_state = CopperState::Begin;
					self.copper_pc += 2;
					inc_pc = false;
				}
			},
			0x5 => /* skip H */
			{
				if x >= ins & 0x3ff {
					self.copper_state = CopperState::Begin;
					self.copper_pc += 2;
					inc_pc = false;
				}
	 
			},
			0x6 => /* update background color and wait next scanline */
			{
				self.r = ((ins & 0xf00) >> 4) as u8;
				self.g = (ins & 0x0f0) as u8;
				self.b = ((ins & 0x00f) << 4) as u8;
				self.hv_wait = y + 1;
				self.copper_state = CopperState::WaitV;
			},
			0x7 => /* V wait N */
			{
				self.hv_wait = y + (ins & 0x3ff);
				self.copper_state = CopperState::WaitV;
	
			},
			0x8 => /* H wait N */
			{
				self.hv_wait = x + (ins & 0x3ff);
				self.copper_state = CopperState::WaitH;
			},
			0x9 => /* write gfx register */
			{

			},
			0xa => /* x pan write */
			{
				self.x_pan = ins & 0x3ff; 
			},
			0xb => /* x pan write and v wait next scanline */
			{
				self.x_pan = ins & 0x3ff; 
				self.hv_wait = y + 1;
				self.copper_state = CopperState::WaitV;
			},
			0xc => /* write background color */ {
				self.r = ((ins & 0xf00) >> 4) as u8;
				self.g = (ins & 0x0f0) as u8;
				self.b = ((ins & 0x00f) << 4) as u8; 
			},
			0xd => /* write alpha */ {
				self.global_alpha = (ins & 0xf) as u8;
			},
			_ => {}
		}

		if inc_pc {
			self.copper_pc += 1;
		}
	}

	pub fn wait_hv(& mut self, wait : u16)
	{
		if wait >= self.hv_wait {
			self.copper_state = CopperState::Execute;
		}
	}
	
	pub fn write_reg(& mut self) -> (bool, u16, u16)
	{
		let reg_data = self.copper_list[(self.copper_pc & 0x1ff) as usize];
		self.copper_state = CopperState::Execute; 
		return (true, self.reg_addr, reg_data);
	}

	pub fn step(& mut self, x : u16, y : u16, _hs: bool, vs: bool) -> 
		(u8 /* BG Color R */, u8 /* G */, u8 /* B */, u16 /* x out */, u16 /* y out */, bool /* register write? */, u16 /* register address */, u16 /* register data */) 
	{


		if vs {

			if self.enable { self.copper_state = CopperState::Begin;}
			self.copper_pc = 0;
		}
		else
		{

			match self.copper_state
			{
				CopperState::Idle => {}, 
				CopperState::Begin => self.begin(),
				CopperState::Execute => self.execute(x, y),
				CopperState::WaitV => self.wait_hv(y),
				CopperState::WaitH => self.wait_hv(x),
				CopperState::WriteReg => {
					let (write_reg, reg, reg_data) = self.write_reg();
					return (self.r, self.g, self.b, x, y, write_reg, reg, reg_data);
				}
			}

		}

		let mut yy = y;

		if self.y_flip_en && y > self.y_flip {
			yy = (self.y_flip - y) + self.y_flip;
		}

		(self.r, self.g, self.b, x + self.x_pan, yy, false, 0, 0)

	}

}

File: slurmemu-ng/src/Copper.cpp
/* vim: set et ts=4 sw=4: */

/*
	slurmemu-ng : Next-Generation SlURM16 Emulator

Copper.cpp : Emulate copper

License: MIT License

Copyright 2023 J.R.Sharp

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

*/

#include "Copper.h"
//#include <stdio.h>

Copper::Copper()    :
    m_enable(false),
    m_copper_pc(0),
    m_state(COPPER_IDLE),
    m_r(255), m_g(0), m_b(0),
    m_regAddr(0),
    m_hvWait(0),
    m_yFlip(0),
    m_yFlipEn(false),
    m_xPan(0),
    m_globalAlpha(255),
    m_alphaOverride(true)
{


}

std::uint16_t Copper::port_op(std::uint16_t port, bool write, std::uint16_t wr_val)
{
    if (write)
    {
        switch (port & 0xfff)
        {
            case 0xd20:
                m_enable = !!(wr_val & 1);
                break;
            case 0xd21:
                m_yFlip = wr_val & 0x1ff;
                m_yFlipEn = !!(wr_val & 0x8000);
                break;
            case 0xd22:
                m_r = ((wr_val & 0xf00) >> 4);
			    m_g = (wr_val & 0x0f0);
				m_b = ((wr_val & 0x00f) << 4);
	            break;
            case 0xd23:
                m_xPan = (wr_val & 0x1ff);
                break;
            case 0xd24:
                m_alphaOverride = !!(wr_val & 0x8000);
                m_globalAlpha = (wr_val & 0xf) << 4;
                break;
            default:
                if (((port & 0xf00) == 0x400) ||
                    ((port & 0xf00) == 0x500)) {
            
                   m_copperList[((port & 0xfff) - 0x400)] = wr_val;
                }
                break;
        }       
    } 
    return 0;
} 

void Copper::execute(int x, int y)
{

    bool inc_pc = true;

    std::uint16_t ins = m_copperList[(m_copper_pc & 0x1ff)];

    switch ((ins & 0xf000) >> 12)
    {
        case 0x1: /* jump */
            inc_pc = false;
		    m_copper_pc = (ins & 0x1ff);
            break;
        case 0x2:   /* V wait */
            m_hvWait = ins & 0x3ff;
			m_state = COPPER_WAITV;
            break;
        case 0x3:   /* H wait */
		    m_hvWait = ins & 0x3ff;
		    m_state = COPPER_WAITH; 
            break;
        case 0x4:   /* skip V */
            if (y >= (ins & 0x3ff)) 
            {
                m_state = COPPER_BEGIN;
                m_copper_pc += 2;
                inc_pc = false;
            }
            break;
        case 0x5:   /* skip H */
            if (x >= (ins & 0x3ff))
            {
                m_state = COPPER_BEGIN;
                m_copper_pc += 2;
                inc_pc = false;
            }    
            break;
        case 0x6:   /* update background color and wait next scanline */
            m_r = (ins & 0xf00) >> 4;
            m_g = (ins & 0x0f0);
            m_b = (ins & 0x00f) << 4;
            m_hvWait = y + 1;
            m_state = COPPER_WAITV;
            break;
        case 0x7:   /* V wait N */
            m_hvWait = y + (ins & 0x3ff);
            m_state = COPPER_WAITV;
            break;
        case 0x8: /* H wait N */
            m_hvWait = x + (ins & 0x3ff);
            m_state = COPPER_WAITH;
            break;
        case 0x9:   /* register write */

            break;
        case 0xa:   /* x pan write */
            m_xPan = ins & 0x3ff;
            break;
        case 0xb:   /* x pan write and v wait next scanline */
            m_xPan = ins & 0x3ff;
            m_hvWait = y + 1;
            m_state = COPPER_WAITV;
            break;
        case 0xc: /* write background color */ 
            m_r = (ins & 0xf00) >> 4;
            m_g = (ins & 0x0f0);
            m_b = (ins & 0x00f) << 4;
            break;
        case 0xd: /* write alpha */ 
			m_globalAlpha = (ins & 0xf) << 4;
            break;
        default: 
            break;
    }

    if (inc_pc)
        m_copper_pc++;

} 

void Copper::begin()
{
    m_state = COPPER_EXECUTE;
}

void Copper::wait_hv(std::uint16_t hv)
{
    /*static std::uint16_t wait = 0;

    if (wait != m_hvWait)
    {
        printf("CPR wait: %d %d\n", m_hvWait, hv);
        wait = m_hvWait;
    }*/

    if (hv >= m_hvWait)
        m_state = COPPER_EXECUTE; 
}
  
void Copper::step(std::uint16_t* mem, uint16_t x1, uint16_t y1, uint16_t x_in, uint16_t y_in, uint16_t& x, uint16_t& y, bool vs)
{

    if (y_in == 0)
    {
        if (m_enable)
        {
            m_state = COPPER_BEGIN;
            m_copper_pc = 0;
        }
        else
        {
            m_state = COPPER_IDLE;
        }
    }
    else
    {
        switch (m_state)
        {
            case COPPER_IDLE: 
                break;
            case COPPER_BEGIN: 
                begin();
                break;
            case COPPER_EXECUTE: 
                execute(x_in, y_in);
                break;
            case COPPER_WAITH:
                wait_hv(x_in);
                break;
            case COPPER_WAITV:
                wait_hv(y_in);
                break;
            case COPPER_WRITEREG:
                break;
        }
    }

    m_xpanTexture[y1] = m_xPan;

    int idx = (x1 + y1*800);
    
    m_copperBG[idx] = m_r + (m_g << 8) + (m_b << 16) + (m_globalAlpha << 24);
   
    m_alphaOverrideTexture[y1][0] = m_alphaOverride ? 1.0 : 0.0;
    m_alphaOverrideTexture[y1][1] = m_globalAlpha / 240.0; 
    
    x = x_in;
    y = y_in;
}

